overlord/devicestate/devicestate.go

```diff
@@ -1411,35 +1457,261 @@ func createRecoverySystemTasks(st *state.State, label string, snapSetupTasks []s
+
+	// LocalSnapPaths is a list of paths to snaps that corresponds to the list
+	// of snap.SideInfo structs in SideInfos. It must be the same length as
+	// LocalSnapSideInfos.
+	LocalSnapPaths []string
+
+	// TestSystem is set to true if the new recovery system should be verified
+	// by rebooting into the new system, prior to marking it as a valid recovery
+	// system. If false, the system will immediately be considered a valid
+	// recovery system.
+	TestSystem bool
+
+	// MarkCurrent is set to true if the new recovery system should be
+	// marked as the current system.
+	MarkCurrent bool
+}
+
+var ErrNoRecoverySystem = errors.New("recovery system does not exist")
+
+// CreateRecoverySystem creates a new recovery system with the given label. See
+// CreateRecoverySystemOptions for details on the options that can be provided.
+func CreateRecoverySystem(st *state.State, label string, opts CreateRecoverySystemOptions) (chg *state.Change, err error) {
 	var seeded bool
-	err := st.Get("seeded", &seeded)
+	err = st.Get("seeded", &seeded)
 	if err != nil && !errors.Is(err, state.ErrNoState) {
 		return nil, err
 	}
 	if !seeded {
 		return nil, fmt.Errorf("cannot create new recovery systems until fully seeded")
 	}
-	chg := st.NewChange("create-recovery-system", fmt.Sprintf("Create new recovery system with label %q", label))
-	ts, err := createRecoverySystemTasks(st, label, nil)
+
+	valsets := snapasserts.NewValidationSets()
+	for _, vs := range opts.ValidationSets {
+		if err := valsets.Add(vs); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := valsets.Conflict(); err != nil {
+		return nil, err
+	}
+
+	revisions, err := valsets.Revisions()
+	if err != nil {
+		return nil, err
+	}
+
+	model, err := findModel(st)
+	if err != nil {
+		return nil, err
+	}
+
+	// check that all snaps from the model are constained by a validation set
+	// and check that all snaps from the model are valid in the validation sets.
+	if err := checkAllSnapsFromModelInValidationSets(model, valsets); err != nil {
+		return nil, err
+	}
+
+	// snaps required by validation sets must also be required by the model
+	if err := checkForRequiredSnapsNotRequiredInModel(model, valsets); err != nil {
+		return nil, err
+	}
+
+	tracker := snap.NewSelfContainedSetPrereqTracker()
+	offline := len(opts.LocalSnapPaths) > 0
+
+	var downloadTSS []*state.TaskSet
+	for _, sn := range model.AllSnaps() {
+		rev := revisions[sn.Name]
+
+		needsInstall, err := snapNeedsInstall(st, sn.Name, rev)
+		if err != nil {
+			return nil, err
+		}
+
+		if !needsInstall {
+			info, err := snapstate.CurrentInfo(st, sn.Name)
+			if err != nil {
+				return nil, err
+			}
+			tracker.Add(info)
+			continue
+		}
+
+		if offline {
+			info, err := offlineSnapInfo(sn, rev, opts)
+			if err != nil {
+				return nil, err
+			}
+			tracker.Add(info)
+			continue
+		}
+
+		const userID = 0
+		ts, info, err := snapstateDownload(context.TODO(), st, sn.Name, dirs.SnapBlobDir, &snapstate.RevisionOptions{
```

Comment:
mmh, downloading things this way one by one, respects validations sets but not old style refresh-control we should leave at a comment about that for now